File: cli/src/services/cost-calculator.ts
import pricingData from '../data/pricing.json' with { type: 'json' }
import type { ModelPricing, PricingMap } from '../types.js'
import { loadCachedPricing } from './pricing-cache.js'

// Bundled pricing is the always-present baseline/floor. If the optional runtime
// refresh has written a cache, overlay it (additive — it only adds/updates models,
// never removes the bundled floor). Computed once at module load.
const pricing: PricingMap = (() => {
  const base: PricingMap = { ...(pricingData as PricingMap) }
  const overlay = loadCachedPricing()
  return overlay ? { ...base, ...overlay } : base
})()

interface CostInput {
  model: string
  inputTokens: number
  outputTokens: number
  cacheReadTokens: number
  cacheWriteTokens: number
  // Bill at the OpenAI priority service tier when the model exposes priority rates.
  priority?: boolean
}

const ZERO_PRICING: ModelPricing = {
  inputPerMillion: 0,
  outputPerMillion: 0,
  cacheReadPerMillion: 0,
  cacheWritePerMillion: 0,
}

// Deterministic normalization: from the most specific model id, strip recognized
// trailing markers one at a time — @YYYYMMDD (vertex), :N (bedrock), -vN, -YYYYMMDD —
// yielding progressively-shorter candidates that are each matched EXACTLY.
// We never scan for arbitrary prefix overlaps: that silently mis-priced new models
// against unrelated siblings (e.g. 'gemini-3-pro' -> 'gemini-3-pro-image-preview').
function candidateKeys(model: string): string[] {
  const out: string[] = []
  const seen = new Set<string>()
  const push = (k: string) => { if (k && !seen.has(k)) { seen.add(k); out.push(k) } }

  push(model)
  let cur = model
  for (let i = 0; i < 8; i++) {
    let next = cur
    if (/@\d{8}$/.test(next)) next = next.replace(/@\d{8}$/, '')
    else if (/:\d+$/.test(next)) next = next.replace(/:\d+$/, '')
    else if (/-v\d+$/.test(next)) next = next.replace(/-v\d+$/, '')
    else if (/-\d{8}$/.test(next)) next = next.replace(/-\d{8}$/, '')
    if (next === cur) break
    push(next)
    cur = next
  }
  return out
}

export interface PricingLookup {
  pricing: ModelPricing
  priced: boolean
}

export function lookupPricing(model: string): PricingLookup {
  for (const key of candidateKeys(model)) {
    const p = pricing[key]
    if (p) return { pricing: p, priced: true }
  }
  // Unknown model: return an explicit unpriced sentinel (cost $0) rather than
  // guessing a neighbor's price. Callers should mark the session unpriced.
  return { pricing: ZERO_PRICING, priced: false }
}

export function isModelPriced(model: string): boolean {
  return lookupPricing(model).priced
}

export interface CostResult {
  costMillicents: number
  priced: boolean
}

export function priceSession(input: CostInput): CostResult {
  const { pricing: p, priced } = lookupPricing(input.model)

  // Tier is decided by the request's total input context (fresh + cached +
  // cache-creation). Matching real provider billing, once it crosses the
  // threshold the *whole* request bills at the above-threshold rates — a step
  // function, not the marginal per-band split CodexBar uses for Claude.
  const inputContext = input.inputTokens + input.cacheReadTokens + input.cacheWriteTokens
  const over = p.thresholdTokens != null && inputContext > p.thresholdTokens

  // Priority rates (when requested and available) take precedence; otherwise the
  // above-threshold rate applies when over the threshold; otherwise the base rate.
  const pick = (base: number, above?: number, priority?: number): number => {
    if (input.priority && priority != null) return priority
    if (over && above != null) return above
    return base
  }

  const inputRate = pick(p.inputPerMillion, p.inputPerMillionAboveThreshold, p.priorityInputPerMillion)
  const outputRate = pick(p.outputPerMillion, p.outputPerMillionAboveThreshold, p.priorityOutputPerMillion)
  const cacheReadRate = pick(p.cacheReadPerMillion, p.cacheReadPerMillionAboveThreshold, p.priorityCacheReadPerMillion)
  // Cache-creation has no priority tier; it only varies by context threshold.
  const cacheWriteRate = over && p.cacheWritePerMillionAboveThreshold != null
    ? p.cacheWritePerMillionAboveThreshold
    : p.cacheWritePerMillion

  const cost =
    (input.inputTokens * inputRate +
      input.outputTokens * outputRate +
      input.cacheReadTokens * cacheReadRate +
      input.cacheWriteTokens * cacheWriteRate) /
    1_000_000

  return { costMillicents: Math.round(cost * 100_000), priced }
}

export function calculateCostMillicents(input: CostInput): number {
  return priceSession(input).costMillicents
}

export function calculateCost(input: CostInput): number {
  return calculateCostMillicents(input) / 100_000
}

File: menubar-app/Sources/Core/CostCalculator.swift
import Foundation

public final class CostCalculator: @unchecked Sendable {
    public static let shared = CostCalculator()

    /// Bump when the lookup/pricing logic changes in a way that should re-cost
    /// already-stored rows. Rows with a lower pricing_version are recomputed once
    /// by SessionsRepo.backfillPricing. v1 = exact-match lookup (no fuzzy fallback).
    public static let currentPricingVersion = 1

    private let pricing: [String: ModelPricing]

    public convenience init() {
        // In a shipped .app read pricing.json from Contents/Resources/
        // via Bundle.main. Only fall back to Bundle.module in dev/test —
        // merely referencing Bundle.module on an end-user Mac triggers
        // SwiftPM's generated accessor which fatalErrors on missing
        // build-dir paths.
        let bundle: Bundle
        if Bundle.main.bundlePath.hasSuffix(".app") {
            bundle = Bundle.main
        } else {
            bundle = Bundle.module
        }
        self.init(bundle: bundle)
    }

    internal init(bundle: Bundle) {
        self.pricing = Self.load(bundle: bundle)
    }

    private static func load(bundle: Bundle) -> [String: ModelPricing] {
        guard let url = bundle.url(forResource: "pricing", withExtension: "json"),
              let data = try? Data(contentsOf: url) else {
            return [:]
        }
        return (try? JSONDecoder().decode([String: ModelPricing].self, from: data)) ?? [:]
    }

    /// Deterministic normalization candidates: from the most specific id, strip
    /// recognized trailing markers one at a time (@YYYYMMDD, :N, -vN, -YYYYMMDD)
    /// and match each EXACTLY. Never a substring scan — that silently mis-priced
    /// new models against unrelated siblings (e.g. gemini-3-pro -> *-image-preview).
    /// Mirrors candidateKeys() in cli/src/services/cost-calculator.ts.
    static func candidateKeys(_ model: String) -> [String] {
        var out: [String] = []
        var seen = Set<String>()
        func push(_ k: String) { if !k.isEmpty && !seen.contains(k) { seen.insert(k); out.append(k) } }
        push(model)
        var cur = model
        for _ in 0..<8 {
            var next = cur
            if let r = next.range(of: #"@\d{8}$"#, options: .regularExpression) { next.removeSubrange(r) }
            else if let r = next.range(of: #":\d+$"#, options: .regularExpression) { next.removeSubrange(r) }
            else if let r = next.range(of: #"-v\d+$"#, options: .regularExpression) { next.removeSubrange(r) }
            else if let r = next.range(of: #"-\d{8}$"#, options: .regularExpression) { next.removeSubrange(r) }
            if next == cur { break }
            push(next)
            cur = next
        }
        return out
    }

    public func lookup(_ model: String) -> ModelPricing? {
        for key in Self.candidateKeys(model) {
            if let p = pricing[key] { return p }
        }
        return nil
    }

    /// Whether the model has a known price. Unknown models are "unpriced" (cost $0
    /// as a placeholder), distinct from a genuine $0 — callers should flag them.
    public func isPriced(_ model: String) -> Bool {
        lookup(model) != nil
    }

    /// Returns cost in millicents (1/1000 of a cent).
    public func cost(
        model: String,
        inputTokens: Int, outputTokens: Int,
        cacheReadTokens: Int = 0, cacheWriteTokens: Int = 0
    ) -> Int {
        guard let price = lookup(model) else { return 0 }
        let dollars =
            Double(inputTokens) / 1_000_000.0 * price.inputPerMillion +
            Double(outputTokens) / 1_000_000.0 * price.outputPerMillion +
            Double(cacheReadTokens) / 1_000_000.0 * price.cacheReadPerMillion +
            Double(cacheWriteTokens) / 1_000_000.0 * price.cacheWritePerMillion
        return Int((dollars * 100_000).rounded())
    }
}
